Pegasus/include/Geometry.hpp

```diff
@@ -171,6 +171,397 @@ class Box : public SimpleShape
 
 namespace intersection
 {
+/** Data structure to store ray factors for Ray collisions */
+struct RayIntersectionFactors
+{
+    double tMin;
+    double tMax;
+};
+
+/**
+* @brief Returns true if a ray and a sphere are intersecting
+* @param[in] raySphere vector from the ray origin to the sphere center
+* @param[in] sphereRadius radius of the sphere
+* @param[in] rayDirection normalized direction vector of the ray
+* @return @c true if there is intersection, @c false otherwise
+*/
+bool CalculateRaySphereIntersection(
+    glm::dvec3 const& raySphere, double sphereRadius, glm::dvec3 const& rayDirection
+);
+
+/**
+* @brief Calculates ray factors for the sphere intersection points
+*
+* Must be called if and only if the ray and the sphere are intersecting
+* @param[in] raySphere vector from the ray to the sphere center
+* @param[in] sphereRadius radius of the sphere
+* @param[in] rayDirection normalized direction vector of the ray
+* @return ray intersection factors
+*/
+RayIntersectionFactors CalculateRaySphereIntersectionFactors(
+    glm::dvec3 const& raySphere, double sphereRadius, glm::dvec3 const& rayDirection
+);
+
+/** Data structure for storing AABB using minimum and maximum points */
+struct AabbExtremalVertices
+{
+    glm::dvec3 minVertex;
+    glm::dvec3 maxVertex;
+};
+
+/**
+* @brief Calculates ray intersection factors for AABB-Ray collision
+* @param[in] boxMinPoint min point of AABB
+* @param[in] boxMaxPoint max point of AABB
+* @param[in] rayDirection normalized direction vector
+* @param[in] rayOrigin ray origin
+* @return ray intersection factors
+*/
+RayIntersectionFactors CalculateRayAabbIntersectionFactors(
+    glm::dvec3 const& boxMinPoint, glm::dvec3 const& boxMaxPoint,
+    glm::dvec3 const& rayDirection, glm::dvec3 const& rayOrigin
+);
+
+/**
+ * @brief Calculates AABB min and max points from the given OBB basis
+ * @param[in] i vector from an orthogonal basis different from j, k
+ * @param[in] j vector from an orthogonal basis different from i, k
+ * @param[in] k vector from an orthogonal basis different from i, j
+ * @return AABB min and max points
+ */
+AabbExtremalVertices MakeExtremalVerticesAabb(
+    glm::dvec3 const& i, glm::dvec3 const& j, glm::dvec3 const& k
+);
+
+/**
+ * @brief Returns true if a ray and an aabb are intersecting
+ * @param[in] tMin min ray intersection factor
+ * @param[in] tMax max ray intersection factor
+ * @return @c true if there is intersection, @c false otherwise
+ */
+bool CalculateRayAabbIntersection(double tMin, double tMax);
+
+/**
+ * @brief Calculates whether two vectors are pointed in the same direction and returns true if they are
+ * @param[in] aVector input vector
+ * @param[in] bVector input vector
+ * @return @c true if dot product is positive @c false otherwise
+ */
+inline bool IsSameDirection(glm::dvec3 const& aVector, glm::dvec3 const& bVector)
```

Comment:
Forgot to mention: this name is really confusing. Probably should use something like `AtAcuteAngle` or something.